pkg/provisioner/ironic/raid.go

```diff
@@ -210,6 +210,12 @@ func BuildRAIDCleanSteps(raidInterface string, target *metal3v1alpha1.RAIDConfig
 	}
 
 	// Hardware RAID
+	// If hardware RAID configuration is nil,
+	// keep old hardware RAID configuration
+	if target == nil || target.HardwareRAIDVolumes == nil {
+		return
+	}
+
```

Comment:
The motivation for that case is unclear to me, and it also seems not correct, being in contrast with this later [check](https://github.com/metal3-io/baremetal-operator/pull/941/files#diff-d3542d6731fdd4d0ef86ed2bc8f7bbad22a4a0af2b8c28d0107835a32f7afce1L235-L237) - which instead appeared to be fine: if the desired raid configuration (`target`) is nil (or `HardwareRAIDVolumes` is nil or empty) then it's just cleared from the server (cleaning up any potentially existing previous configuration) (an this could also cover the case of multiple loops on the `Preparing` state, where maybe something meanwhile has changed). Also the software raid part follows a similar logic.

So, assuming that the current master implementation looks fine, could you please elaborate with more details the use case that you think it's missing (unfortunately also the PR description isn't very informative for that) ?

Reply:
In the current master implementation, if the target node does not have a RAID controller, ironic will return clean failed because the `delete_configuration` clean step failed.

Reply:
Therefore, our goal is:

- If hardwareRAIDVolume = nil
Keep actual RAID configuration

- If hardwareRAIDVolume = []
Clear actual RAID configuration

- Other
Apply RAID configuration to node

Reply:
I just found that there is a redundant `if` statement in the place that you mentioned, I dropped it.

Reply:
Additional explaintion:

This is the missing case:
User apply a new bmh without `raid` field, and at the same time, the target node doesn't have RAID controller.

In above case, when entering BuildRAIDCleanSteps(), the values of target and actual are:
```go
target = {
    hardwareRAIDVolume = nil
    softwareRAIDVolume = []
}

actual = nil
```

Because of above values of target and actual, BMO will append `delete_configuration` into clean steps, then due to there is no RAID controller, an error will happen, which leads to reentry the preparing status (infinity loop).